File: mac/Core/Network/ControlMessages.swift
import Foundation

public struct ScreenAnnouncement: Codable, Equatable {
    public let peerId: String
    public let screens: [AnnouncedScreen]

    public init(peerId: String, screens: [AnnouncedScreen]) {
        self.peerId = peerId
        self.screens = screens
    }

    private enum CodingKeys: String, CodingKey {
        case peerId = "peer_id"
        case screens
    }
}

public struct AnnouncedScreen: Codable, Equatable {
    public let id: Int
    public let x: Int
    public let y: Int
    public let w: Int
    public let h: Int
    public let scale: Double
    public let primary: Bool

    public init(id: Int, x: Int, y: Int, w: Int, h: Int, scale: Double, primary: Bool) {
        self.id = id
        self.x = x
        self.y = y
        self.w = w
        self.h = h
        self.scale = scale
        self.primary = primary
    }

    public init(display: DisplayInfo) {
        self.init(
            id: display.screenId,
            x: display.x,
            y: display.y,
            w: display.width,
            h: display.height,
            scale: display.scale,
            primary: display.isPrimary
        )
    }

    public func displayInfo(peerId: String) -> DisplayInfo {
        DisplayInfo(peerId: peerId, screenId: id, x: x, y: y, width: w, height: h, scale: scale, isPrimary: primary)
    }
}

public struct LayoutSyncMessage: Codable, Equatable {
    public let layout: [ScreenRect]

    public init(layout: [ScreenRect]) {
        self.layout = layout
    }
}

public struct HelloMessage: Codable, Equatable {
    public let name: String
    public let os: String
    public let ver: String
    public let screens: [AnnouncedScreen]

    public init(name: String, os: String, ver: String, screens: [AnnouncedScreen]) {
        self.name = name
        self.os = os
        self.ver = ver
        self.screens = screens
    }
}

File: windows/Core/Network/ControlMessages.cs
using System.Text.Json.Serialization;
using BarelyReal.Core.Layout;

namespace BarelyReal.Core.Network;

public sealed record ScreenAnnouncement(
    [property: JsonPropertyName("peer_id")] string PeerId,
    [property: JsonPropertyName("screens")] IReadOnlyList<AnnouncedScreen> Screens);

public sealed record AnnouncedScreen(
    [property: JsonPropertyName("id")] int Id,
    [property: JsonPropertyName("x")] int X,
    [property: JsonPropertyName("y")] int Y,
    [property: JsonPropertyName("w")] int W,
    [property: JsonPropertyName("h")] int H,
    [property: JsonPropertyName("scale")] double Scale,
    [property: JsonPropertyName("primary")] bool Primary)
{
    public static AnnouncedScreen FromDisplay(DisplayInfo display) =>
        new(display.ScreenId, display.X, display.Y, display.Width, display.Height, display.Scale, display.IsPrimary);

    public DisplayInfo ToDisplayInfo(string peerId) =>
        new(peerId, Id, X, Y, W, H, Scale, Primary);
}

public sealed record LayoutSyncMessage(
    [property: JsonPropertyName("layout")] IReadOnlyList<ScreenRect> Layout);

public sealed record HelloMessage(
    [property: JsonPropertyName("name")] string Name,
    [property: JsonPropertyName("os")] string Os,
    [property: JsonPropertyName("ver")] string Ver,
    [property: JsonPropertyName("screens")] IReadOnlyList<AnnouncedScreen> Screens);
